lib/types/message.js

```diff
@@ -1,347 +1,191 @@
- '-------------------------------------------+--------------------------------------------'
-                                             |
-                                      Annotated Message
- </pre>
- *
- * The message _may_ contain the sections above, and application data _may_ be repeated, as follows:
- *
- * * Zero or one {@link Header} sections.
- * * Zero or one {@link DeliveryAnnotations} sections.
- * * Zero or one {@link Annotations} sections.
- * * Zero or one {@link Properties} sections.
- * * Zero or one {@link ApplicationProperties} sections.
- * * The body consists of either: one or more {@link Data} sections, one or more {@link AMQPSequence} sections,
- *      or a single {@link AMQPValue} section.
- * * Zero or one {@link Footer} sections.
- *
- * @param contents
- * @param body
- * @constructor
- */
-function Message(contents, body) {
-  contents = contents || {};
-  u.assignDefined(this, contents, {
-    header: u.coerce(contents.header, Header),
-    deliveryAnnotations: u.coerce(contents.deliveryAnnotations, DeliveryAnnotations),
-    annotations: u.coerce(contents.annotations, Annotations),
-    properties: u.coerce(contents.properties, Properties),
-    applicationProperties: u.coerce(contents.applicationProperties, ApplicationProperties),
-    footer: u.coerce(contents.footer, Footer)
-  });
-
-  this.body = contents.body || body;
-}
+module.exports.encodeMessage = function(message, buffer) {
+  var _keys = Object.keys(sectionByProperty), _len = _keys.length;
+  for (var i = 0; i < _len; ++i) {
+    var property = _keys[i], definition = sectionByProperty[property];
+    if (message.hasOwnProperty(property)) {
+      var section = message[property];
+      if (definition.encode && typeof definition.encode === 'function') {
+        section = definition.encode(section);
+      }
+
+      var value = definition.hasOwnProperty('source') ?
+        new ForcedType(definition.source, section) : section;
+      var described = new DescribedType(definition.descriptor, value);
+      codec.encode(described, buffer);
+    }
+  }
 
-Message.prototype.encode = function(codec, buf) {
-  if (this.header) codec.encode(this.header, buf);
-  if (this.deliveryAnnotations) codec.encode(this.deliveryAnnotations, buf);
-  if (this.annotations) codec.encode(this.annotations, buf);
-  if (this.properties) codec.encode(this.properties, buf);
-  if (this.applicationProperties) codec.encode(this.applicationProperties, buf);
-  if (this.body instanceof Buffer) {
-    codec.encode(new Data(this.body), buf);
-  } else if (Array.isArray(this.body)) {
-    codec.encode(new AMQPSequence(this.body), buf);
-  } else {
-    codec.encode(new AMQPValue(this.body), buf);
+  if (!!message.body) {
+    if (message.body instanceof Buffer) {
+      codec.encode(new DescribedType(0x75, message.body), buffer);  // Data
+    } else if (Array.isArray(message.body)) {
+      codec.encode(new DescribedType(0x76, message.body), buffer);  // AMQPSequence
```

Comment:
do we want to deal with arrays of buffers as multiple Data segments here? arrays of arrays as multiple sequences? We never did before, so I'm fine leaving it out, but with your refactor it makes it _so much easier_.